Api improvements and bug fixes

budgetdevv opened the pull request:
Library was shipped in a broken state, hence this PR exists. Changes I've made include - 

- Fixed issues related to stream not closing properly ( `Z_STREAM_END` is never returned )

- Got rid of `Stream` API, and replaced it with `Span<byte>` ones

- Instead of pinning `zstream`, an uninitialized instance is allocated on stack, and its address is taken

- Removed redundant throws ( `InflateEnd()` / `DeflateEnd()` would throw anyway )

- Renamed methods for consistency

- `MemoryZlib.Compress()` / `MemoryZlib.Decompress()` now returns `uint` denoting number of unprocessed bytes. A return value of 0 would mean that the destination buffer was sufficiently allocated

- `MemoryZlib.Decompress()` now includes an `out uint WrittenBytes` parameter denoting bytes written to the destination buffer. This is useful if the destination buffer is over-allocated

- Unit Tests to test Compression / Decompression

- Tons of minor fixes I forgot about 👀

budgetdevv committed:
Initial commit

Signed-off-by: BudgetDevv <budgetdevv@gmail.com>

budgetdevv committed:
Added out parameter BytesWritten for Decompress(), denoting the number of bytes written to destination buffer

Signed-off-by: BudgetDevv <budgetdevv@gmail.com>

budgetdevv committed:
Clean-up and signoff

Signed-off-by: BudgetDevv <budgetdevv@gmail.com>

AraHaan commented on src/ZlibSharp/Directory.Build.targets:
```diff
@@ -3,6 +3,8 @@
     <Using Include="System.Runtime.InteropServices" />
     <Using Include="System.Runtime.Serialization" />
+    <Using Include="System.Runtime.CompilerServices"/>
+    
```
Remove extra blank line.

budgetdevv replied:
;o

AraHaan commented on src/ZlibSharp/ZlibSharp.csproj:
```diff
@@ -1 +1,6 @@
-<Project Sdk="Microsoft.NET.Sdk" />
+<Project Sdk="Microsoft.NET.Sdk">
+    <ItemGroup>
```
The package should go in to Directory.Packages.props.

budgetdevv replied:
Oops, forgot to remove that

AraHaan commented on tests/ZlibSharp.Tests.csproj:
```diff
@@ -1 +1 @@
-<Project Sdk="Microsoft.NET.Sdk" />
+<Project Sdk="Microsoft.NET.Sdk" />
```
```suggestion
<Project Sdk="Microsoft.NET.Sdk" />

```

AraHaan commented on src/ZlibSharp/ZlibSharp.csproj:
```diff
@@ -1 +1,6 @@
-<Project Sdk="Microsoft.NET.Sdk" />
+<Project Sdk="Microsoft.NET.Sdk">
+    <ItemGroup>
+        <PackageReference Include="FluentAssertions" Version="6.2.0" />
+    </ItemGroup>
+</Project>
```
```suggestion
<Project Sdk="Microsoft.NET.Sdk" />
```

AraHaan commented on src/ZlibSharp/Directory.Build.targets:
```diff
@@ -3,6 +3,8 @@
     <Using Include="System.Runtime.InteropServices" />
     <Using Include="System.Runtime.Serialization" />
+    <Using Include="System.Runtime.CompilerServices"/>
+    
```
```suggestion
```

AraHaan commented on src/ZlibSharp/ZlibSharp/MemoryZlib.cs:
```diff
@@ -8,264 +8,101 @@ namespace ZlibSharp;
-    public static void Decompress(byte[] inData, Stream outStream)
+    
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static uint Compress(Span<byte> Source, Span<byte> Dest, ZlibCompressionLevel CompressionLevel = ZlibCompressionLevel.DefaultCompression)
```
After thinking about this perhaps the source spans could be implicitly converted to ``ReadOnlySpan``s and nuke the byte[] overloads so users do not get too big of a breakage.

budgetdevv replied:
You can't write to a readonlyspan, though

budgetdevv replied:
I think an implicit cast to Span<byte> is possible, though. Imma try to refactor the API

budgetdevv replied:
Just realized I can't get rid of the .AsSpan() calls, since it would cause a recursive call into the byte[] API

AraHaan replied:
Not if you have the span ones call directly into what they need to call into.

AraHaan committed:
Fix tests, added back native library resolver for MacOS and linux, other span based api changes.

This should cleanup the api as well as fix running the code on non-Windows systems as well as make tests work cross platform.

As for Windows, I have no idea how to resolve possible issues with loading 32 bit or 64 bit zlib dlls accidentally within the wrong bitness other than bundling the windows versions of the dlls with the library.

Signed-off-by: AraHaan <seandhunt_7@yahoo.com>

budgetdevv commented:
Oops, please ignore the last push

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+internal static unsafe class ZlibHelper
 {
-    internal static unsafe bool InitializeInflate(ref ZStream zs)
+    private static void InitializeInflate(ZStream* StreamPtr)
```
```suggestion
    private static void InitializeInflate(ZStream* streamPtr)
```

AraHaan replied:
Move private functions after internal ones.

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-        fixed (ZStream* pzs = &zs)
+        var Result = UnsafeNativeMethods.inflateInit_(StreamPtr, UnsafeNativeMethods.zlibVersion(), sizeof(ZStream));
+        
+        if (Result != ZlibResult.Ok)
```
```suggestion
        var result = UnsafeNativeMethods.inflateInit_(streamPtr, UnsafeNativeMethods.zlibVersion(), sizeof(ZStream));
        if (result != ZlibResult.Ok)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-            }
-
-            return true;
+            throw new NotPackableException($"{nameof(InitializeInflate)} failed - ({Result}) {Marshal.PtrToStringUTF8((nint) StreamPtr->msg)}");        
```
```suggestion
            throw new NotPackableException($"{nameof(InitializeInflate)} failed - ({result}) {Marshal.PtrToStringUTF8((nint)streamPtr->msg)}");        
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
     }
 
-    internal static unsafe bool InitializeDeflate(ref ZStream zs, ZlibCompressionLevel compressionLevel)
+    private static void InitializeDeflate(ZStream* StreamPtr, ZlibCompressionLevel CompressionLevel)
```
```suggestion
    private static void InitializeDeflate(ZStream* streamPtr, ZlibCompressionLevel CompressionLevel)
```

AraHaan replied:
Move private functions after internal ones.

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-        fixed (ZStream* pzs = &zs)
+        var Result = UnsafeNativeMethods.deflateInit_(StreamPtr, CompressionLevel, UnsafeNativeMethods.zlibVersion(), sizeof(ZStream));
+        
+        if (Result != ZlibResult.Ok)
```
```suggestion
        var result = UnsafeNativeMethods.deflateInit_(streamPtr, CompressionLevel, UnsafeNativeMethods.zlibVersion(), sizeof(ZStream));
        if (result != ZlibResult.Ok)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-            }
-
-            return true;
+            throw new NotPackableException($"{nameof(InitializeDeflate)} failed - ({Result}) {Marshal.PtrToStringUTF8((nint) StreamPtr->msg)}");        
```
```suggestion
            throw new NotPackableException($"{nameof(InitializeDeflate)} failed - ({result}) {Marshal.PtrToStringUTF8((nint)streamPtr->msg)}");        
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
     }
 
-    internal static unsafe bool FinalizeInflate(ref ZStream zs)
+    internal static void InflateEnd(ZStream* StreamPtr)
```
```suggestion
    internal static void InflateEnd(ZStream* streamPtr)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-        fixed (ZStream* pzs = &zs)
+        var Result = UnsafeNativeMethods.inflateEnd(StreamPtr);
+        
+        if (Result != ZlibResult.Ok)
```
```suggestion
        var result = UnsafeNativeMethods.inflateEnd(streamPtr);
        if (result != ZlibResult.Ok)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+        if (Result != ZlibResult.Ok)
         {
-            return UnsafeNativeMethods.inflateEnd(pzs) == ZlibResult.Ok;
+            throw new NotPackableException($"{nameof(InflateEnd)} failed - ({Result}) {Marshal.PtrToStringUTF8((nint) StreamPtr->msg)}");
```
```suggestion
            throw new NotPackableException($"{nameof(InflateEnd)} failed - ({result}) {Marshal.PtrToStringUTF8((nint)streamPtr->msg)}");
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
     }
 
-    internal static unsafe bool FinalizeDeflate(ref ZStream zs)
+    private static void DeflateEnd(ZStream* StreamPtr)
```
```suggestion
    private static void DeflateEnd(ZStream* streamPtr)
```

AraHaan replied:
Move private functions after internal ones.

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+    private static void DeflateEnd(ZStream* StreamPtr)
     {
-        fixed (ZStream* pzs = &zs)
+        var Result = UnsafeNativeMethods.deflateEnd(StreamPtr);
+        
+        if (Result != ZlibResult.Ok)
```
```suggestion
        var result = UnsafeNativeMethods.deflateEnd(streamPtr);
        if (result != ZlibResult.Ok)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+        if (Result != ZlibResult.Ok)
         {
-            return UnsafeNativeMethods.deflateEnd(pzs) == ZlibResult.Ok;
+            throw new NotPackableException($"{nameof(DeflateEnd)} failed - ({Result}) {Marshal.PtrToStringUTF8((nint) StreamPtr->msg)}");
```
```suggestion
            throw new NotPackableException($"{nameof(DeflateEnd)} failed - ({result}) {Marshal.PtrToStringUTF8((nint)streamPtr->msg)}");
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
     }
-
-    internal static unsafe ZlibResult Decompress(ref ZStream zs, ref byte[] dest, ref byte[] source, ZlibFlushStrategy flush)
+    internal static uint Compress(Span<byte> Source, Span<byte> Dest, ZlibCompressionLevel CompressionLevel, out uint Adler32)
```
```suggestion
    internal static uint Compress(Span<byte> source, Span<byte> dest, ZlibCompressionLevel compressionLevel, out uint adler32)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+    }
 
-        return ret is ZlibResult.Ok or ZlibResult.StreamEnd;
+    private static uint GetAdler32(ZStream* StreamPtr)
```
Make this expression bodied.

AraHaan replied:
```suggestion
    private static uint GetAdler32(ZStream* streamPtr)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-        return ret is ZlibResult.Ok or ZlibResult.StreamEnd;
+    private static uint GetAdler32(ZStream* StreamPtr)
+    {
+        return (uint) (StreamPtr->adler.Value & 0xffff);
```
```suggestion
        return (uint)(streamPtr->adler.Value & 0xffff);
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
         {
-            // error?
-        }
+            InitializeDeflate(StreamPtr, CompressionLevel);
+            
+            StreamPtr->next_in = SourcePtr;
+            StreamPtr->avail_in = (uint) Source.Length;
+            StreamPtr->next_out = DestPtr;
+            StreamPtr->avail_out = (uint) Dest.Length;
```
```suggestion
            InitializeDeflate(streamPtr, compressionLevel);
            streamPtr->next_in = sourcePtr;
            streamPtr->avail_in = (uint)source.Length;
            streamPtr->next_out = destPtr;
            streamPtr->avail_out = (uint)dest.Length;
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-        {
-            throw new NotUnpackableException($"Exception during zlib decompression: ({ret}) {Marshal.PtrToStringUTF8((nint)zs.msg)}");
-        }
+            ZlibResult Result;
+
+            while (UnsafeNativeMethods.deflate(StreamPtr, ZlibFlushStrategy.NoFlush) == ZlibResult.Ok)
+            {
+                if (StreamPtr->avail_in == 0)
+                {
+                    UnsafeNativeMethods.deflate(StreamPtr, ZlibFlushStrategy.Finish);
+                }
+            };
+
+            Adler32 = GetAdler32(StreamPtr);
```
```suggestion
            ZlibResult result;
            while (UnsafeNativeMethods.deflate(streamPtr, ZlibFlushStrategy.NoFlush) == ZlibResult.Ok)
            {
                if (streamPtr->avail_in == 0)
                {
                    UnsafeNativeMethods.deflate(streamPtr, ZlibFlushStrategy.Finish);
                }
            };

            Adler32 = GetAdler32(streamPtr);
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+            Adler32 = GetAdler32(StreamPtr);
 
-        return ret;
+            DeflateEnd(StreamPtr);
+
+            return (uint) Stream.total_out.Value;
+        }
```
```suggestion
            DeflateEnd(streamPtr);
            return (uint)stream.total_out.Value;
        }
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
 
-    internal static unsafe bool Compress(ref ZStream zs, ref byte[] dest, ref byte[] source, ZlibFlushStrategy flush)
+    //Decompress returns avail_in, allowing users to reallocate and continue decompressing remaining data
+    //should Dest buffer be under-allocated
+
+    internal static uint Decompress(Span<byte> Source, Span<byte> Dest, out uint BytesWritten, out uint Adler32)
```
```suggestion
    //should Dest buffer be under-allocated
    internal static uint Decompress(Span<byte> source, Span<byte> dest, out uint bytesWritten, out uint adler32)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-            zs.avail_out = (uint)dest.Length;
-            ret = UnsafeNativeMethods.deflate(pzs, flush);
-        }
+        ZStream Stream;
 
```
```suggestion
        ZStream stream;
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-            zs.avail_out = (uint)dest.Length;
-            ret = UnsafeNativeMethods.inflate(pzs, flush);
-        }
+        ZStream Stream;
 
```
```suggestion
        ZStream stream;
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+        ZStream Stream;
 
-        if ((ulong)dest.LongLength < zs.total_out.Value)
+        var StreamPtr = &Stream;
+        
+        //We skipped initialization
+        StreamPtr->zalloc = null;
+        StreamPtr->zfree = null;
+        
+        fixed (byte* SourcePtr = Source)
+        fixed (byte* DestPtr = Dest)
```
```suggestion
        var streamPtr = &stream;

        //We skipped initialization
        streamPtr->zalloc = null;
        streamPtr->zfree = null;
        fixed (byte* sourcePtr = source)
        fixed (byte* destPtr = dest)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
-        {
-            // error?
-        }
+        var StreamPtr = &Stream;
```
```suggestion
        var streamPtr = &stream;
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+        var StreamPtr = &Stream;
 
-        if (ret is not ZlibResult.Ok and not ZlibResult.StreamEnd)
+        //We skipped initialization
+        StreamPtr->zalloc = null;
+        StreamPtr->zfree = null;
+        
+        fixed (byte* SourcePtr = Source)
+        fixed (byte* DestPtr = Dest)
```
```suggestion
        // We skipped initialization
        streamPtr->zalloc = null;
        streamPtr->zfree = null;
        fixed (byte* sourcePtr = source)
        fixed (byte* destPtr = dest)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,128 @@
+        fixed (byte* DestPtr = Dest)
         {
-            throw new NotPackableException($"deflating: ({ret}) {Marshal.PtrToStringUTF8((nint)zs.msg)}");
+            InitializeInflate(StreamPtr);
+            
+            StreamPtr->next_in = SourcePtr;
+            StreamPtr->avail_in = (uint) Source.Length;
+            StreamPtr->next_out = DestPtr;
+            StreamPtr->avail_out = (uint) Dest.Length;
+
+            while (UnsafeNativeMethods.inflate(StreamPtr, ZlibFlushStrategy.NoFlush) == ZlibResult.Ok)
+            {
+                if (StreamPtr->avail_in == 0)
+                {
+                    UnsafeNativeMethods.inflate(StreamPtr, ZlibFlushStrategy.Finish);
+                }
+            }
+
+            BytesWritten = (uint) StreamPtr->total_out.Value;
+            
+            Adler32 = GetAdler32(StreamPtr);
+
+            InflateEnd(StreamPtr);
+
+            return StreamPtr->avail_in;
```
```suggestion
            InitializeInflate(streamPtr);
            streamPtr->next_in = sourcePtr;
            streamPtr->avail_in = (uint)source.Length;
            streamPtr->next_out = destPtr;
            streamPtr->avail_out = (uint)dest.Length;
            while (UnsafeNativeMethods.inflate(streamPtr, ZlibFlushStrategy.NoFlush) == ZlibResult.Ok)
            {
                if (streamPtr->avail_in == 0)
                {
                    UnsafeNativeMethods.inflate(streamPtr, ZlibFlushStrategy.Finish);
                }
            }

            BytesWritten = (uint)streamPtr->total_out.Value;
            Adler32 = GetAdler32(streamPtr);
            InflateEnd(streamPtr);
            return streamPtr->avail_in;
```

AraHaan commented on tests/Test.cs:
```diff
@@ -0,0 +1,81 @@
+using System.IO.Compression;
+using System.Runtime.InteropServices;
+using FluentAssertions;
+using Xunit;
+using Xunit.Abstractions;
```
Add these to Directory.Build.targets for the test project.

AraHaan commented on tests/Test.cs:
```diff
@@ -0,0 +1,81 @@
+
+        BytesWritten.Should().Be(SourceLength);
+    }
+}
```
```suggestion
}

```

AraHaan commented on tests/Test.cs:
```diff
@@ -0,0 +1,81 @@
+    [Fact]
+    public void DecompressionToOverAllocatedBufferShouldHaveBytesWrittenEqualToSourceStringLength()
+    {
+        const uint OversizeBy = 69;
+
+        var SourceLength = (uint) SourceString.Length;
+        
+        var OversizedDestBuffer = new byte[SourceLength + OversizeBy];
+        
+        MemoryZlib.Decompress(SourceStringCompressed, OversizedDestBuffer, out var BytesWritten).Should().Be(0);
+
+        BytesWritten.Should().Be(SourceLength);
```
```suggestion
        const uint oversizeBy = 69;
        var sourceLength = (uint)sourceString.Length;
        var oversizedDestBuffer = new byte[sourceLength + oversizeBy];
        MemoryZlib.Decompress(sourceStringCompressed, oversizedDestBuffer, out var bytesWritten).Should().Be(0);
        bytesWritten.Should().Be(sourceLength);
```

AraHaan commented on tests/Test.cs:
```diff
@@ -0,0 +1,81 @@
+    [Fact]
+    public void CompressionToUnderAllocatedBufferReturnsNonZeroValue()
+    {
+        const int UndersizedBufferLength = 69;
+
+        UndersizedBufferLength.Should().BeLessThan(LengthOfCompressed);
+        
+        var UndersizedDestBuffer = new byte[UndersizedBufferLength];
+        
+        MemoryZlib.Compress(SourceStringCompressed, UndersizedDestBuffer).Should().NotBe(0);
```
```suggestion
        const int undersizedBufferLength = 69;
        undersizedBufferLength.Should().BeLessThan(lengthOfCompressed);
        var undersizedDestBuffer = new byte[undersizedBufferLength];
        MemoryZlib.Compress(sourceStringCompressed, undersizedDestBuffer).Should().NotBe(0);
```

AraHaan commented on tests/Test.cs:
```diff
@@ -0,0 +1,81 @@
+
+public class Test
+{
+    private readonly ITestOutputHelper FakeConsole;
+    
+    private readonly byte[] SourceString, SourceStringCompressed, SourceBuffer;
+
+    private readonly int LengthOfCompressed;
```
```suggestion
    private readonly ITestOutputHelper fakeConsole;
    private readonly byte[] sourceString, sourceStringCompressed, sourceBuffer;
    private readonly int lengthOfCompressed;
```

AraHaan commented on tests/Test.cs:
```diff
@@ -0,0 +1,81 @@
+
+    private readonly int LengthOfCompressed;
+    
+    public Test(ITestOutputHelper fakeConsole)
+    {
+        FakeConsole = fakeConsole;
+        
+        SourceString = File.ReadAllBytes("SourceText.txt");
+
+        var DestBuffer = new byte[SourceString.Length];
+        
+        LengthOfCompressed = (int) MemoryZlib.Compress(SourceString, DestBuffer);
+
+        SourceStringCompressed = new byte[LengthOfCompressed];
+        
+        DestBuffer.AsSpan(0, LengthOfCompressed).CopyTo(SourceStringCompressed);
+
+        SourceBuffer = new byte[SourceString.Length];
+    }
```
```suggestion
    public Test(ITestOutputHelper fakeConsole)
    {
        this.fakeConsole = fakeConsole;
        this.sourceString = File.ReadAllBytes("SourceText.txt");
        var destBuffer = new byte[sourceString.Length];
        this.lengthOfCompressed = (int)MemoryZlib.Compress(sourceString, destBuffer);
        this.sourceStringCompressed = new byte[lengthOfCompressed];
        this.destBuffer.AsSpan(0, this.lengthOfCompressed).CopyTo(this.sourceStringCompressed);
        this.sourceBuffer = new byte[this.sourceString.Length];
    }
```

AraHaan commented on tests/Test.cs:
```diff
@@ -0,0 +1,81 @@
+    [Fact]
+    public void DecompressionWorks()
+    {
+        MemoryZlib.Decompress(SourceStringCompressed, SourceBuffer, out _).Should().Be(0);
+
+        SourceBuffer.Should().Equal(SourceString);
```
```suggestion
        MemoryZlib.Decompress(sourceStringCompressed, sourceBuffer, out _).Should().Be(0);
        sourceBuffer.Should().Equal(sourceString);
```

AraHaan commented on tests/Test.cs:
```diff
@@ -0,0 +1,81 @@
+    [Fact]
+    public void DecompressionToUnderAllocatedBufferReturnsNonZeroValue()
+    {
+        const int UndersizedBufferLength = 69;
+
+        UndersizedBufferLength.Should().BeLessThan(SourceBuffer.Length);
+        
+        var UndersizedDestBuffer = new byte[UndersizedBufferLength];
+        
+        MemoryZlib.Decompress(SourceStringCompressed, UndersizedDestBuffer, out var BytesWritten).Should().NotBe(0);
+
+        BytesWritten.Should().Be(UndersizedBufferLength);
```
```suggestion
        const int undersizedBufferLength = 69;
        undersizedBufferLength.Should().BeLessThan(sourceBuffer.Length);
        var undersizedDestBuffer = new byte[undersizedBufferLength];
        MemoryZlib.Decompress(sourceStringCompressed, undersizedDestBuffer, out var bytesWritten).Should().NotBe(0);
        bytesWritten.Should().Be(undersizedBufferLength);
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,111 @@
+    private static uint GetAdler32(ZStream* streamPtr)
+    {
+        return (uint)(streamPtr->adler.Value & 0xffff);
     }
```
```suggestion
    private static uint GetAdler32(ZStream* streamPtr)
        => (uint)(streamPtr->adler.Value & 0xffff);
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,111 @@
     }
 
-    internal static unsafe bool InitializeDeflate(ref ZStream zs, ZlibCompressionLevel compressionLevel)
+    private static void InitializeDeflate(ZStream* streamPtr, ZlibCompressionLevel CompressionLevel)
```
```suggestion
    private static void InitializeDeflate(ZStream* streamPtr, ZlibCompressionLevel compressionLevel)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,111 @@
+    private static void InitializeDeflate(ZStream* streamPtr, ZlibCompressionLevel CompressionLevel)
     {
-        fixed (ZStream* pzs = &zs)
+        var result = UnsafeNativeMethods.deflateInit_(streamPtr, CompressionLevel, UnsafeNativeMethods.zlibVersion(), sizeof(ZStream));
```
```suggestion
        var result = UnsafeNativeMethods.deflateInit_(streamPtr, compressionLevel, UnsafeNativeMethods.zlibVersion(), sizeof(ZStream));
```

AraHaan commented on src/ZlibSharp/ZlibSharp/ZlibHelper.cs:
```diff
@@ -5,108 +5,111 @@
     }
 
-    internal static unsafe bool FinalizeInflate(ref ZStream zs)
+    internal static void InflateEnd(ZStream* streamPtr)
```
```suggestion
    private static void InflateEnd(ZStream* streamPtr)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/MemoryZlib.cs:
```diff
@@ -281,16 +118,23 @@ public static string ZlibVersion()
+    }
+    
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static ulong ZlibGetAdler32(Span<byte> Data)
```
Cant this be ``ReadOnlySpan`` or does zlib's adler32 function modify the incoming pointer?

AraHaan replied:
```suggestion
    public static ulong ZlibGetAdler32(ReadOnlySpan<byte> data)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/MemoryZlib.cs:
```diff
@@ -281,16 +118,23 @@ public static string ZlibVersion()
     /// <returns>The Adler32 hash of the input data.</returns>
-    public static unsafe ulong ZlibGetAdler32(byte[] data)
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static ulong ZlibGetAdler32(byte[] Data)
+    {
+        return ZlibGetAdler32(Data.AsSpan());
+    }
+    
```
```suggestion
```

Cant we remove the array version so that way the implicit conversion to ReadOnlySpan can be used for the Span version of this?

AraHaan commented on src/ZlibSharp/ZlibSharp/MemoryZlib.cs:
```diff
@@ -281,16 +118,23 @@ public static string ZlibVersion()
+    public static ulong ZlibGetAdler32(Span<byte> Data)
     {
-        fixed (byte* pdata = data)
+        fixed (byte* DataPtr = Data)
```
```suggestion
        fixed (byte* dataPtr = data)
```

AraHaan commented on src/ZlibSharp/ZlibSharp/MemoryZlib.cs:
```diff
@@ -281,16 +118,23 @@ public static string ZlibVersion()
                 UnsafeNativeMethods.adler32(0L, null, 0),
-                pdata,
-                (uint)data.Length);
+                DataPtr,
+                (uint) Data.Length);
```
```suggestion
                dataPtr,
                (uint)data.Length);
```

AraHaan commented:
Refreshing Codacy now.

AraHaan commented:
@budgetdevv Codacy does not like that ``bytesWritten`` is an output parameter, I feel the best option would be to directly return an readonly struct to hold the information they might need?

AraHaan committed:
Make MemoryZlib compress and decompress methods return readonly struct instances.

Signed-off-by: AraHaan <seandhunt_7@yahoo.com>

AraHaan committed:
Fix build on MacOS.

AraHaan committed:
Add exception with custom message when zlib cant be found.

AraHaan committed:
Updated tests, and other fixes.

AraHaan committed:
Added crc32 test.

AraHaan committed:
Fix loading on Linux and MacOS.

AraHaan commented:
This PR now waits on https://github.com/madler/zlib/pull/732.

AraHaan committed:
Added additional test.

AraHaan committed:
Fixed endline normalization on text txt files.

AraHaan committed:
Update .gitattributes

AraHaan committed:
Changes.

AraHaan committed:
Testing.

AraHaan committed:
Use native zlib from nuget packages.
